tldw_Server_API/tests/Research/test_research_discovery_catalog.py

```
@@ -0,0 +1,115 @@
+    from tldw_Server_API.app.core.Research.discovery.catalog import (
+        ResearchSourceCatalog,
+        default_source_catalog,
+    )
+
+    base_source = default_source_catalog().get_source("openalex")
+    duplicate_entries = [
+        base_source,
+        replace(base_source, display_name="Duplicate OpenAlex"),
+    ]
+
+    with pytest.raises(ValueError, match="^duplicate_source_id:openalex$"):
```

Ruff (0.15.17)

[warning] Pattern passed to `match=` contains metacharacters but is neither escaped nor raw

(RUF043)